src/helpers/buffers/unordered_receiver.rs

```diff
@@ -203,30 +215,78 @@ where
 
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
+    ///
+    /// ## Errors
+    /// If buffer capacity is not enough to read `M` and the underlying stream does not have
+    /// more data. This may lead to different behavior depending on the order of issued reads. See
+    /// [`read_order`] test for an example.
+    ///
+    /// [`read_order`]: test::read_order
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+        let message = self.spare.read();
+        let next = if let Some(m) = message {
+            // this check exists to make sure the inner stream is eventually moved to
+            // the closed state. We don't want to poll it too often, but we also need to know
+            // when it is done and `UnorderedReceiver` can be dropped.
             self.wake_next();
-            return Poll::Ready(Ok(m));
-        }
-
-        loop {
-            match self.stream.as_mut().poll_next(cx) {
-                Poll::Pending => {
-                    return Poll::Pending;
-                }
-                Poll::Ready(Some(b)) => {
-                    if let Some(m) = self.spare.extend(b.as_ref()) {
+            Ok(m)
+        } else {
+            loop {
+                if let Some(bytes) = ready!(self.stream.as_mut().poll_next(cx)) {
+                    if let Some(m) = self.spare.extend(bytes.as_ref()) {
                         self.wake_next();
-                        return Poll::Ready(Ok(m));
+                        break Ok(m);
                     }
+                } else {
+                    self.spare.replace(&[]);
+                    self.close();
+                    break Err(Error::EndOfStream {
+                        record_id: RecordId::from(self.next),
+                    });
+                }
+            }
+        };
+
+        if next.is_ok() && self.spare.is_empty() {
+            // we don't want to be woken up here, control loop is driven by the client.
+            // They decide when they want the next message and must issue a `poll` for it.
+
+            // TODO: https://github.com/rust-lang/rust/issues/98286
+            let mut cx = Context::from_waker(noop_waker_ref());
+            match self.stream.as_mut().poll_next(&mut cx) {
```

Comment:
@martinthomson We discussed that a real waker should be used here if there are no active wakers in the receiver and it is the last record. When the underlying stream returns `Poll::Ready` everything is fine, however there seems to be a subtle violation of `Stream` contract - we already have an item in our hand to return (line 271 `next`) but if underlying stream is not ready, the executor is not be obligated to poll us again, because we returned `Poll::Ready`. 

I could probably make it work by holding onto `next` and returning `Poll::Pending`, but that's not great in case when we are racing with receivers, i.e. it is not the end of the stream yet.

Trying to think if there is a good solution here.

Reply:
Hmm, that doesn't sound great.

We could do as you say and always be one item ahead.  That would ensure that we deliver the last item and close at the same time.  But I think that the core problem is a lack of anything interested in the close.

Would it make sense for us to have the context/gateway infrastructure be interested in the close of each receiver?  That is, for each receiver that we register, we have that hooked into the context, with a future that resolves when the receiver closed.  Before we close a query, we would await those futures to ensure that they were driven to completion (you could also spawn a task for that purpose that is in turn awaited before the transaction resolves).